server/src/main/java/org/opensearch/index/query/TemplateQueryBuilder.java

```diff
@@ -0,0 +1,158 @@
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        throw new IllegalStateException("Template query should run with a ml_inference request processor");
+    }
+
+    @Override
+    protected boolean doEquals(TemplateQueryBuilder other) {
+        return Objects.equals(this.content, other.content);
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(content);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    /**
+     * Gets the content of this template query.
+     *
+     * @return The template content as a map.
+     */
+    public Map<String, Object> getContent() {
+        return content;
+    }
+
+    /**
+     * Rewrites the template query by substituting variables from the context.
+     *
+     * @param queryShardContext The context for query rewriting.
+     * @return A rewritten QueryBuilder.
+     * @throws IOException If there's an error during rewriting.
+     */
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
+        if (!(queryShardContext instanceof QueryCoordinatorContext)) {
+            throw new IllegalStateException("Template query needs to be resolved with variables from search processors.");
+        }
+
+        QueryCoordinatorContext queryCoordinateContext = (QueryCoordinatorContext) queryShardContext;
+
+        Map<String, Object> contextVariables = queryCoordinateContext.getContextVariables();
+        String queryString;
+
+        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+            builder.map(this.content);
+            queryString = builder.toString();
+        }
+
+        // Convert Map<String, Object> to Map<String, String> with proper JSON escaping
+        Map<String, String> variablesMap = contextVariables.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> {
+            try {
+                return JsonXContent.contentBuilder().value(entry.getValue()).toString();
+            } catch (IOException e) {
+                throw new RuntimeException("Error converting contextVariables to JSON string", e);
+            }
+        }));
+
+        StringSubstitutor substitutor = new StringSubstitutor(variablesMap).setVariablePrefix("\"${").setVariableSuffix("}\"");
+        String newQueryContent = substitutor.replace(queryString);
+
+        try {
+            XContentParser parser = XContentType.JSON.xContent()
+                .createParser(queryShardContext.getXContentRegistry(), LoggingDeprecationHandler.INSTANCE, newQueryContent);
+
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+
+            QueryBuilder newQueryBuilder = parseInnerQueryBuilder(parser);
+
+            return newQueryBuilder;
+
+        } catch (Exception e) {
+            throw new IOException("Failed to rewrite template query: " + newQueryContent, e);
```

Comment:
Why IOException, and not something of RuntimeException like IllegalArgumentException or similar?

Reply:
agree, it make more sense to be IllegalArgumentException